datastore/bolt/api_key.go

```diff
@@ -2,54 +2,249 @@ package bolt
-	return nil, nil
+func (a *apiKeyRepo) FindAPIKeyByMaskID(ctx context.Context, maskID string) (*datastore.APIKey, error) {
+	var apiKeys []datastore.APIKey = make([]datastore.APIKey, 0, 1)
+	var apiKey datastore.APIKey
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		c := b.Cursor()
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			err := json.Unmarshal(v, &apiKey)
+			if err != nil {
+				return err
+			}
+
+			if apiKey.MaskID == maskID {
+				apiKeys = append(apiKeys, apiKey)
+				break
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil && len(apiKeys) == 0 {
+		return &apiKey, err
+	}
+
+	if err == nil && len(apiKeys) == 0 {
+		return &apiKey, datastore.ErrAPIKeyNotFound
+	}
+
+	return &apiKeys[0], err
 }
 
-func (e *apiKeyRepo) FindAPIKeyByMaskID(ctx context.Context, maskID string) (*datastore.APIKey, error) {
-	return nil, nil
+func (a *apiKeyRepo) RevokeAPIKeys(ctx context.Context, uids []string) error {
+	err := a.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		for _, uid := range uids {
+			err := b.Delete([]byte(uid))
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	return err
 }
 
-func (e *apiKeyRepo) RevokeAPIKeys(ctx context.Context, uids []string) error {
-	return nil
+func (a *apiKeyRepo) FindAPIKeyByHash(ctx context.Context, hash string) (*datastore.APIKey, error) {
+	var apiKeys []datastore.APIKey = make([]datastore.APIKey, 0, 1)
+	var apiKey datastore.APIKey
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		c := b.Cursor()
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			err := json.Unmarshal(v, &apiKey)
+			if err != nil {
+				return err
+			}
+
+			if apiKey.Hash == hash {
+				apiKeys = append(apiKeys, apiKey)
+				break
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil && len(apiKeys) == 0 {
+		return &apiKey, err
+	}
+
+	if err == nil && len(apiKeys) == 0 {
+		return &apiKey, datastore.ErrAPIKeyNotFound
+	}
+
+	return &apiKeys[0], err
 }
 
-func (e *apiKeyRepo) FindAPIKeyByHash(ctx context.Context, hash string) (*datastore.APIKey, error) {
-	return nil, nil
+func (a *apiKeyRepo) LoadAPIKeysPaged(ctx context.Context, pageable *datastore.Pageable) ([]datastore.APIKey, datastore.PaginationData, error) {
+	var apiKeys []datastore.APIKey = make([]datastore.APIKey, 0)
+
+	page := pageable.Page
+	prevPage := pageable.Page
+	perPage := pageable.PerPage
+	data := datastore.PaginationData{}
+
+	if pageable.Page < 1 {
+		page = 1
+	}
+
+	if pageable.PerPage < 1 {
+		perPage = 10
+	}
+
+	if page < 1 {
+		prevPage = 1
+	} else {
+		prevPage = page - 1
+	}
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		c := b.Cursor()
+		i := 1
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			if v == nil {
+				continue
+			}
+
+			if i > perPage*prevPage && i <= perPage*page {
```

Comment:
i tihnk storing the bounds makes sense here, instead of having computed on every iteration,
```go
lowerBound := perPage*prevPage
upperBound := perPage*page
```